ARCs/arc-0027.md

```diff
@@ -0,0 +1,1611 @@
+  - **MUST** have a maximum length of 16 as this is the maximum amount of grouped (atomic) transactions that are allowed by the Algorand protocol
+* `reference`:
+  - **MUST** be `arc0027:sign_txns:request`
+
+**Response Payload**
+
+```json
+{
+  "title": "Sign Transactions Response Payload",
+  "type": "object",
+  "properties": {
+    "error": {
+      "type": "object",
+      "description": "Error payload associated with the response"
+    },
+    "id": {
+      "type": "string",
+      "description": "A globally unique identifier for the message"
+    },
+    "reference": {
+      "type": "string",
+      "description": "Identifies the purpose of the message"
+    },
+    "requestId": {
+      "type": "string",
+      "description": "The ID of the request message"
+    },
+    "result": {
+      "type": "object",
+      "properties": {
+        "providerId": {
+          "type": "string",
+          "description": "A unique identifier for the provider"
+        },
+        "stxns": {
+          "type": "array",
+          "description": "A list of base64 encoded signed transactions that is ready to be posted to the network",
+          "items": {
+            "anyOf": [
+              {
+                "type": "string"
+              },
+              {
+                "type": "null"
+              }
+            ]
+          }
+        }
+      },
+      "required": [
+        "providerId",
+        "stxns"
+      ]
+    }
+  },
+  "required": [
+    "id",
+    "reference",
+    "requestId"
+  ]
+}
+```
+where:
+* `error.code`:
+  - **MAY** be [4001 `MethodCanceledError`](#4001-methodcancelederror) if the request was canceled by the user
+  - **MAY** be [4002 `MethodTimedOutError`](#4002-methodtimedouterror) if the request timed out
+  - **MAY** be [4003 `MethodNotSupportedError`](#4003-methodnotsupportederror) if the method is not supported by the provider
+  - **MAY** be [4004 `NetworkNotSupportedError`](#4004-networknotsupportederror) if the `genesisHash` provided in the transactions to be signed is not supported by the provider
+  - **MAY** be [4100 `UnauthorizedSignerError`](#4100-unauthorizedsignererror) if a signer in the request is not authorized by the provider
+  - **MAY** be [4200 `InvalidInputError`](#4200-invalidinputerror) if computed group ID for the `txns` in the request does not match the assigned group ID
+  - **MAY** be [4201 `InvalidGroupIdError`](#4201-invalidgroupiderror) if the `txns` in the request is malformed
+* `result.providerId`:
+  - **MUST** be a <a href="https://www.rfc-editor.org/rfc/rfc4122">UUIDv4</a> compliant string
+  - **MUST** uniquely identify the provider
+* `result.stxns`:
+  - **MAY** have an item that is a base64 encoding of a signed transaction
+  - **MAY** have an item that is null if the transaction at the same index in the request's `txns` list was not signed by the provider
+  - **MAY** be empty
+* `reference`:
+  - **MUST** be `arc0027:sign_txns:response`
+
+#### Sign Bytes
```

Comment:
I'm a fan of us supporting signing for multiple purposes and arbitrary data signing. But `signBytes`: 
- seems to give freedom to the app to define what `purposes` this can be used for with a simple string. Seems that it can be gamed to trick users to sign things that they shouldn't, even transactions. 
- Doesn't consider multiple encoding schemes
- Validation of the content to be signed should be matched against a known schema? 

That's one of the things i'm trying to address in [ARC-52#SignMetadata](https://github.com/algorandfoundation/ARCs/pull/239/files#diff-5f82fbf5efe9e776afbc7c317caf254ed75d9428879851529dc85f34b7338345R41) is to enforce the context of what we are signing and validate it quickly from a list of possible schema templates (i.e authentication requests, messages, identity), so that the user doesn't have to understand msgpack encoding or CBOR and the wallet can still validate what is signing. 

It's a longer conversation though.

Reply:
I wholeheartedly agree. I have been playing with this with [Algorand Provider](https://algorand-provider.agoralabs.sh/) + Kibisis and, if I recall, Kibisis displays the data if it is:
* A valid UTF-8 string
* A parseable JWT

For example, for JWTs, if it is correctly parsed it shows this screen:
<img alt="Screen of Kibisis JWT signing" src="https://github.com/algorandfoundation/ARCs/assets/6410708/21ef1386-17ef-443c-b3c1-d84d9c9aa775" width="300px"/>

It also shows some warnings if the JWT has expired or the subject/sub of the JWT does not match the account to sign. 

As you can workout, I am a fan of JWTs :sweat_smile: 

(It was actually the only reason I built Kibisis as I wanted to implement, for the then, app I was building to use user JWT authentication using an Algorand account)

But you are right, the schema of this function requires more work as it is not fit for consumption.